metagraph/plugins/numpy/types.py

```diff
@@ -230,17 +230,42 @@ def __init__(self, data, *, mask=None, node_ids=None):
                 len(id2pos) == len(data), f"node_ids must be the same length as data"
             )
 
-    def __getitem__(self, node_id):
+    def __len__(self):
+        return self.num_nodes
+
+    def _get_single_item(self, node_id):
         if self.mask is not None:
-            if self.mask[node_id]:
+            if not self.mask[node_id]:
                 raise ValueError(f"node {node_id} is not in the NodeMap")
         elif self.id2pos is not None:
             if node_id not in self.id2pos:
                 raise ValueError(f"node {node_id} is not in the NodeMap")
-            pos = self.id2pos[node_id]
-            return self.value[pos]
+            node_id = self.id2pos[node_id]
         return self.value[node_id]
 
+    def _get_multiple_items(self, node_ids: np.ndarray):
+        if self.mask is not None:
+            if (node_ids >= len(self.mask)).any() or not self.mask[node_ids].all():
+                out_of_bounds_ids = node_ids[node_ids >= len(self.mask)]
+                mask_missing_ids = np.flatnonzero(~self.mask[node_ids])
+                missing_nodes = list(out_of_bounds_ids) + list(mask_missing_ids)
+                raise ValueError(f"nodes {missing_nodes} are not in the NodeMap")
+        elif self.id2pos is not None:
+            missing_ids = [
+                node_id for node_id in node_ids if node_id not in self.id2pos
+            ]
+            if missing_ids:
+                raise ValueError(f"nodes {missing_ids} are not in the NodeMap")
+            node_ids = [self.id2pos[node_id] for node_id in node_ids]
+        return self.value[node_ids]
+
+    def __getitem__(self, key):
+        return (
+            self._get_multiple_items(key)
+            if isinstance(key, np.ndarray)
```

Comment:
If you're going to make this handle single ints and ndarrays, it should also handle tuples/lists of indices for completeness. Simply check for `hasattr(key, '__len__')` and then wrap it in `np.array(key)` if it isn't already an ndarray.

Reply:
https://github.com/metagraph-dev/metagraph/pull/133/commits/7d5693ea7003e52b39ab71f1d795198acfb099cb implements this suggestion. 

This is slightly different than what NumPy supports for the tuple case, e.g. 
* This works with `NumpyNodeMap`
```
nm = NumpyNodeMap(...)
nm[(3,1,0)]
```
* The behavior for NumPy is different:
```
>>> np.arange(10)[(9,3,0)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: too many indices for array: array is 1-dimensional, but 3 were indexed
>>> 
```

@jim22k , does this sound reasonable?

Reply:
Yes, this is still reasonable. Numpy doesn't handle the tuple case because `nm[(3, 1, 0)]` is identical to `nm[3, 1, 0]` and ndarrays can be multidimensional. But NodeMaps can't have more than one dimension, so we can correctly interpret a tuple as a list of indices to fetch.